### PR TITLE
exporter broken tracking

### PR DESCRIPTION
**Pull request by jluebbe**
**Description**
When the exporter is unable to start or stop subprocesses for resources, it's better to handle that explicitly rather than wait for errors when using them.

**Checklist**
- [x] Tests for the feature 
- [x] PR has been tested

### CODE REVIEW

File: labgrid/remote/exporter.py [hunk 1]
```diff
@@ -48,6 +67,22 @@ def __attrs_post_init__(self):
         for key in self.local_params:
             del self.params[key]
         self.start_params = None
+        self._broken = None
+
+    # is something criticial failed for an export, we can mark it as
```
Comment (Bastian-Krause) [inline labgrid/remote/exporter.py]:
```suggestion
    # if something criticial failed for an export, we can mark it as
```

File: labgrid/remote/exporter.py [hunk 2]
```diff
@@ -48,6 +67,22 @@ def __attrs_post_init__(self):
...
+    # permanently broken
+    @property
+    def broken(self):
+        return self._broken
+
+    @broken.setter
+    def broken(self, reason):
+        assert self._broken is None
+        assert type(reason) == str
+        assert len(reason) > 0
```
Comment (Bastian-Krause) [inline labgrid/remote/exporter.py]:
```suggestion
        assert reason
```
Should be enough at this point.

File: labgrid/remote/exporter.py [hunk 3]
```diff
@@ -48,6 +67,22 @@ def __attrs_post_init__(self):
...
+        self._broken = reason
+        self.data['acquired'] = '<broken>'
```
Comment (Bastian-Krause) [inline labgrid/remote/exporter.py]: I guess this needs a little documentation somewhere.

File: labgrid/remote/exporter.py [hunk 4]
```diff
@@ -90,38 +139,48 @@ def poll(self):
...
+            params['extra']['broken'] = self.broken
         if self.params != params:
             self.data['params'].update(params)  # pylint: disable=unsubscriptable-object
             dirty = True
 
         return dirty
 
     def acquire(self, *args, **kwargs):  # pylint: disable=arguments-differ
+        if self.broken:
+            raise BrokenResourceError("cannot acquire broken resource (original reason): %s", self.broken)
```
Comment (Bastian-Krause) [inline labgrid/remote/exporter.py]: Why do you use the % style formatting here?

File: labgrid/remote/exporter.py [hunk 5]
```diff
@@ -90,38 +139,48 @@ def poll(self):
...
         super().acquire(*args, **kwargs)
         self.poll()
 
     def release(self, *args, **kwargs):  # pylint: disable=arguments-differ
+        if self.broken:
+            raise BrokenResourceError("cannot release broken resource (original reason): %s", self.broken)
```
Comment (Bastian-Krause) [inline labgrid/remote/exporter.py]: And here?

File: tests/conftest.py [hunk 6]
```diff
@@ -117,7 +117,11 @@ def exporter(tmpdir, crossbar):
         """
     Testport:
         NetworkSerialPort:
-          {host: 'localhost', port: 4000}
+          host: 'localhost'
+          port: 4000
+    Broken:
+        RawSerialPort:
+          port: 'none'
```
Comment (Bastian-Krause) [inline tests/conftest.py]: Just to make sure I understand this: This lets `assert start_params['path'].startswith('/dev/')` raise an AssertionError. This is then caught and `broken` is set to 'start failed'. Correct?

Comment (jluebbe) [reply tests/conftest.py]: Yes, this was the easiest way to support a test for this code.

File: labgrid/remote/exporter.py (hunk 3 above)
Comment (jluebbe) [reply labgrid/remote/exporter.py]: I've added a comment here.

File: labgrid/remote/exporter.py [hunk 7]
```diff
@@ -35,6 +35,17 @@ class BrokenResourceError(ExporterError):
...
+def log_subprocess_kernel_stack(logger, child):
+    if child.poll() is not None:  # nothing to check if no longer running
+        return
+    try:
+        with open('/proc/{}/stack'.format(child.pid), 'r') as f:
+            stack = f.read()
+            stack = stack.strip()
+    except PermissionError:
+        return
+    logger.info("current kernel stack of {} is:\n".format(child.args)+stack)
```
Comment (Bastian-Krause) [inline labgrid/remote/exporter.py]: Nit: % style formatting should be used in logging functions.

Comment (jluebbe) [reply labgrid/remote/exporter.py]: Yes, but this is very far from performance critical code. ;)

Comment (Bastian-Krause) [reply labgrid/remote/exporter.py]: Agreed, but this will be another pylint warning.

File: labgrid/remote/exporter.py [hunk 8]
```diff
@@ -27,6 +27,25 @@
...
+def log_subprocess_kernel_stack(logger, child):
+    if child.poll() is not None:  # nothing to check if no longer running
+        return
+    try:
+        with open('/proc/{}/stack'.format(child.pid), 'r') as f:
+            stack = f.read()
+            stack = stack.strip()
+    except PermissionError:
+        return
+    logger.info("current kernel stack of %s is:\n%s" % (child.args, stack))
```
Comment (Bastian-Krause) [inline labgrid/remote/exporter.py]:
```suggestion
    logger.info("current kernel stack of %s is:\n%s", child.args, stack)
```

Comment (jluebbe) [reply labgrid/remote/exporter.py]: Thanks, updated again. :man_facepalming:

**Pull request merged**